Automattic/woocommerce-payments
# Add state handling for WCPay UPE method activation/checkout

Fixes #3077

#### Changes proposed in this Pull Request



This PR provides the merchants to see the approval statuses of their UPE payment methods in their 
payment methods settings page and the UPE onboarding workflow. The changes proposed:

- The users won't be allowed to enable a payment method which has an "inactive" status on the Stripe's account settings.
- If there are previously enabled "inactive" payment methods, the merchant will see a notice pill, telling the merchant to "Contact WooCommerce Support" to get the "inactive" status fixed.
- If there is a pending approval for some payment methods, the merchant will see the "Pending Approval" notice pill, but they will be allowed to enable those. 
- The above items are also true for the onboarding workflow.
- On the checkout page, only "active" payment methods will be shown to the user.
- If there are some connection issues between Stripe and WCPay, the user will always be able to see the "card" payment method on checkout, which may result with an error. But it seems a better option than showing an empty form on the checkout page.

The visual updates are:

![image](https://user-images.githubusercontent.com/3295/137336313-45232549-7ff3-4ac4-bd30-f45b1ea7c69b.png)

![image](https://user-images.githubusercontent.com/3295/137336385-d05e6926-8aa8-434f-ab50-23dd46fc5731.png)

![image](https://user-images.githubusercontent.com/3295/137336723-c4059192-a89d-4ff5-aea0-d0cfcd06dedd.png)

![image](https://user-images.githubusercontent.com/3295/137336794-dd1dbda0-35fd-41cd-af66-001363661519.png)








#### Testing instructions

##### To test it without using a Stripe account:

Return manual statuses by editing `get_upe_enabled_payment_method_statuses` function and returning your settings like this instead:

```php
    return [
		'giropay_payments'    => 'pending_approval',
		'sofort_payments'     => 'pending_verification',
		'sepa_debit_payments' => 'inactive',
		'card_payments'       => 'active',
	];
```

Location: 

https://github.com/Automattic/woocommerce-payments/blob/3d767c1ad4b6472c86f9507f00538f4bbe95eaac/includes/class-wc-payment-gateway-wcpay.php#L2348-L2358

##### To test it using a Stripe account:

1. Onboard to Stripe with a live account, but don't fill your business URL, add a product description instead
2. All UPE payment methods should be having the inactive state except the card payment.
3. Check if you can enable the inactive payment methods in your WC Payments settings page.
4. You can enable the payment methods by switching to a test account
5. Return back to the live account and see if the payment methods in the checkout form contains the inactive payment methods.

---

You can use the Stripe account `acct_1Ji1DW2EGoaSBhgm` for testing this, you need to have database access to your local server and modify the `wcpay_accounts` table to link your setup to this account. Also you need to have your live Stripe keys set in your local server's `secrets.php` file to get it working.

More information about how to get this working: paJDYF-1Ys-p2

-------------------


**Post merge**
